File: algorithms/javaalgo/src/test/java/leetcode/number1to100/number41to50/number45/JumpGame2Test.java
package leetcode.number1to100.number41to50.number45;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import org.junit.jupiter.api.Test;

public class JumpGame2Test {

    @Test
    void test() {
        int[] nums1 = {2, 3, 1, 1, 4};
        int[] nums2 = {2, 3, 0, 1, 4};

        assertAll(
            () -> assertThat(jump(nums1)).isEqualTo(2),
            () -> assertThat(jump(nums2)).isEqualTo(2)
        );
    }

    public int jump(int[] nums) {
        int destination = nums.length - 1;
        int stepCount = 0;
        int lastJumpMax = 0;
        int currentJumpMax = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            currentJumpMax = Math.max(currentJumpMax, i + nums[i]);
            if (i == lastJumpMax) {
                stepCount++;
                lastJumpMax = currentJumpMax;
                if (currentJumpMax >= destination) {
                    return stepCount;
                }
            }
        }
        return stepCount;
    }
}

File: algorithms/javascriptalgo/leetcode/number41-50/number45/jumpGame2.test.js
test("test", () => {
  const nums1 = [2, 3, 1, 1, 4];
  const nums2 = [2, 3, 0, 1, 4];

  expect(jump(nums1)).toEqual(2);
  expect(jump(nums2)).toEqual(2);
});

function jump(nums) {
  const end = nums.length - 1;
  let answer = 0;
  let lastJump = 0;
  let curJump = 0;

  for (let i = 0; i < nums.length - 1; i++) {
    curJump = Math.max(curJump, i + nums[i]);
    if (i === lastJump) {
      answer++
      lastJump = curJump;

      if (curJump >= end) {
        return answer;
      }
    }
  }
  return answer;
}
